aff4/aff4_map.cc

```diff
@@ -152,11 +152,39 @@ std::string AFF4Map::Read(size_t length) {
         {
             std::string data = target_stream->Read(length_to_read_in_target);
             if (data.size() < length_to_read_in_target) {
-                resolver->logger->error(
-                    "Map target {} can not produced required {} bytes. Null padding "
-                    "{} bytes!", target_stream->urn.SerializeToString(),
-                    length_to_read_in_target, length_to_read_in_target - data.size());
+                // Failed to read some portion of memory. On Windows platforms, this is usually
+                // due to Virtual Secure Mode (VSM) memory. Re-read memory in smaller units,
+                // while leaving the unreadable regions null-padded.
+                resolver->logger->info(
+                    "Map target {} can not produced required {} bytes at offset 0x{:x}. Got {} bytes. Will re-read one page at a time.",
+                    target_stream->urn.SerializeToString(),
+                    length_to_read_in_target, offset_in_target, data.size());
+
+                // Reset target_strem back to original position, and then re-read one page at a time.
+                target_stream->Seek(offset_in_target, SEEK_SET);
+
+                // Grow the data buffer to the expected size, filled to the end with null bytes.
                 data.resize(length_to_read_in_target, 0);
+                const char* buffer = data.data();
+
+                size_t reread_total = 0;
+                while (reread_total < length_to_read_in_target) {
+                    size_t reread_want = std::min((size_t)(length_to_read_in_target - reread_total), (size_t)4096); // should be PAGE_SIZE
```

Comment:
I think this 4096 can be taken into the class members. This code would be useful for e.g. imaging from a disk where blocksize is 512.

Reply:
I've created a member variable `max_reread_size` to hold this value with 4d98ff1a0bfe84872d2ac3bb408a559ff33f7b96. It defaults to 4096.